### PR TITLE
Unify method in which file content is read across all resources

### PR DESCRIPTION
**Pull request by eramoto**
There are the similar issues as PR #2302. Almost resources return false positives when a file does not exist or is not read.

### CODE REVIEW

File: lib/utils/read_file.rb
Comment (adamleff) [inline lib/utils/read_file.rb]: Our other utilities are named as nouns... ObjectTraverser, CommentParser... while yours is named like a verb/action. Can I suggest changing this to `FileReader` instead (and alter the file name as well to `file_reader.rb`)?

**Review by adamleff — CHANGES_REQUESTED** [review]
@eramoto This is a supremely great improvement. Excellent use of a mixin to remove repetitive code and make it easier for future resource authors to read file content.

I'd like to request a name change of the module to make it consistent with our other mixins. Otherwise, I think this is fantastic.

**Comment by eramoto** [comment]
Hi @adamleff ,

Thank you for your comments. And I apologize for a lot of notifications on slack at yesterday.
Now I am testing this patch set against test/functional/inspec_vendor_test.rb with Travis CI on my account. Please wait for a while.

**Comment by arlimus** [comment]
Amazing work to unify this @eramoto thank you so much!!

**Comment by eramoto** [comment]
Hi @adamleff @arlimus ,

Since in `test/functional/inspec_vendor_test.rb` it is not known exactly why the profile `ssh-baseline` can not find `/etc/ssh/sshd_config` at Travis CI despite the existence of it [1][2], I left the part of old verification into `lib/resources/ssh_conf.rb`.

Please cloud you merge this PR since I will try to figure out on another PR/issue.

[1] : https://travis-ci.org/eramoto/inspec/jobs/313317863
[2] : The following is a snippet from [1]. (It seems sshd_config exists when using `p Dir.entries("/etc/ssh")` in [1]).
```
# Running:

..........................................................................................................["..", "ssh_host_ecdsa_key", "ssh_import_id", "ssh_host_dsa_key.pub", "moduli", "sshd_config", ".", "ssh_host_rsa_key", "ssh_host_ed25519_key.pub", "ssh_host_dsa_key", "ssh_host_ed25519_key", "ssh_config", "ssh_host_rsa_key.pub", "ssh_host_ecdsa_key.pub"]
["..", "ssh_host_ecdsa_key", "ssh_import_id", "ssh_host_dsa_key.pub", "moduli", "sshd_config", ".", "ssh_host_rsa_key", "ssh_host_ed25519_key.pub", "ssh_host_dsa_key", "ssh_host_ed25519_key", "ssh_config", "ssh_host_rsa_key.pub", "ssh_host_ecdsa_key.pub"]
FF....................................

Finished in 99.405121s, 1.4486 runs/s, 5.7442 assertions/s.

  1) Failure:
example inheritance profile#test_0004_ensure json/check command do not fetch remote profiles if vendored [/home/travis/build/eramoto/inspec/test/functional/inspec_vendor_test.rb:63]:
--- expected
+++ actual
@@ -1 +1,31 @@
-""
+"/home/travis/build/eramoto/inspec/lib/utils/file_reader.rb:11:in `read_file_content': Can't find file: /etc/ssh/sshd_config (Inspec::Exceptions::ResourceSkipped)
+\tfrom /home/travis/build/eramoto/inspec/lib/resources/ssh_conf.rb:65:in `read_content'
+\tfrom /home/travis/build/eramoto/inspec/lib/resources/ssh_conf.rb:70:in `read_params'
+\tfrom /home/travis/build/eramoto/inspec/lib/resources/ssh_conf.rb:49:in `method_missing'
+\tfrom ssh-baseline-master/controls/sshd_spec.rb:461:in `block in load_with_context'
```

File: lib/utils/file_reader.rb
Comment (adamleff) [inline lib/utils/file_reader.rb]: This feels like something that should be left up to each resource. What if a file is expected to be empty?

Comment (eramoto) [reply lib/utils/file_reader.rb]: I thought that users expected that any description exists in a file if they specified its path. But they are permitted to read it even if it is empty so they may expected that. Thus I will fix above.

Comment (adamleff) [reply lib/utils/file_reader.rb]: One suggestion: leave this here but make it optional. The caller can specify an option, like so:

`read_file_contents(‘/tmp/empty’, allow_empty: true)`

Comment (adamleff) [reply lib/utils/file_reader.rb]: Let me know if you like that idea and if you need help implementing it! 🙂

Comment (eramoto) [reply lib/utils/file_reader.rb]: I apologize that I was unaware of your helpful advice.
I implemented to split and wrap it but I will rework to implement the caller having the optional argument.

Comment (eramoto) [reply lib/utils/file_reader.rb]: I changed `read_file_content` as described later.
The optional argument is set at false as default since I think `allow_empty: true` should be explicitly specified in case allowing to read a empty file.

```Ruby
--- a/lib/utils/file_reader.rb
+++ b/lib/utils/file_reader.rb
@@ -2,7 +2,7 @@
 # author: ERAMOTO Masaya

 module FileReader
-  def read_file_content(path)
+  def read_file_content(path, allow_empty = false)
     # these are currently ResourceSkipped to maintain consistency with the resource
     # pre-refactor (which used skip_resource). These should likely be changed to
     # ResourceFailed during a major version bump.
@@ -15,7 +15,8 @@ module FileReader
     if raw_content.nil?
       raise Inspec::Exceptions::ResourceSkipped, "Can't read file: #{path}"
     end
-    if raw_content.empty?
+
+    if !allow_empty && raw_content.empty?
       raise Inspec::Exceptions::ResourceSkipped, "File is empty: #{path}"
     end

```

By the way, this PR does not allow dh_params and x509_certificate resources to read a empty file since they always raise a exception if a file is empty.

**Comment by jquick** [comment]
Hey @eramoto ! Sorry for the hold on this. With inspec 2.0 a lot has changed. Do you mind rebasing this PR off master and we can give it another once over and pick it up?

**Comment by eramoto** [comment]
Hi @jquick ,

For the current master branch (whose HEAD is v2.0.45 or later)
and/or the 1-stable branch (which is that I guess for v1.51.21 or later),
I don't mind rebasing my branch or creating a new PR.

Could you tell me a target branch you wish?
* If the current master branch, I will rebase my branch at this PR.
* If the 1-stable branch, I will create a new PR or, change the target branch of this PR and rebase my branch.


BTW, My local branch has already been rebased on v2.0.37 and included a change that removes the workaround for [the failure of sshd_config on Travis CI](https://travis-ci.org/eramoto/inspec/jobs/313317863).
Since it has occured due to a bug of the sshd sample test set, I created [the PR #100 for dev-sec/ssh-baseline](https://github.com/dev-sec/ssh-baseline/pull/100) to fix it.

**Comment by eramoto** [comment]
Hi @jquick ,

I rebased this PR on v2.0.46 and it is blocked by [PR #100 for dev-sec/ssh-baseline](https://github.com/dev-sec/ssh-baseline/pull/100).
If the PR is merged, the build job xx.3 on Travis CI will succeed and we will be able to merge this PR.

**Comment by jquick** [comment]
Thanks @eramoto !

**Comment by jquick** [comment]
@eramoto - the profile change was merged, can you force out again to start a new batch of tests?

**Comment by eramoto** [comment]
Now all checks including the Travis CI build jobs have passed, since the PR for dev-sec/ssh-baseline was  merged and the problem build job was rerun.
Thank you for rerun of it instead of me.

File: lib/resources/auditd_conf.rb
Comment (jquick) [inline lib/resources/auditd_conf.rb]: May just want to use `@content` here so we don't have to call down again. Another option is to just remove `@content` from the init block and let it be read here.

File: lib/resources/bond.rb
Comment (jquick) [inline lib/resources/bond.rb]: We should be able to ditch this line as it will be set in init.

File: lib/resources/inetd_conf.rb
Comment (jquick) [inline lib/resources/inetd_conf.rb]: Same here, lets just use `@content` in this method as we already have the info. Or remove the `@content` from the init.

File: lib/resources/limits_conf.rb
Comment (jquick) [inline lib/resources/limits_conf.rb]: Same here.

File: lib/resources/login_def.rb
Comment (jquick) [inline lib/resources/login_def.rb]: Same here.

File: lib/resources/ntp_conf.rb
Comment (jquick) [inline lib/resources/ntp_conf.rb]: Same here.

**Review by jquick — CHANGES_REQUESTED** [review]
Hey @eramoto ! Fantastic work here. Just a few notes.

File: lib/resources/auditd_conf.rb
Comment (eramoto) [reply lib/resources/auditd_conf.rb]: Hi @jquick , thanks for pointing them out.

When they do not call the read_file_content method in the init block, i.e. removing it from their init blocks, they do not skip the test and do report it as failure if a target file is not found since their to_s methods seem to interrupt the exception from methods.

Thus I will remove the extra call and use `@content` of the init block in the method.

**Review by jquick — APPROVED** [review]
Thanks @eramoto !

**Comment by jquick** [comment]
@adamleff - You mind giving this another view?

**Comment by eramoto** [comment]
Hi @jquick ,

Could you please merge this PR since the merging policy seems to require at least 1 approval and you have approved? If it is the sum of approval, where approval is +1 and non-approval is -1, and @adamleff does not have time to review, could you assign anyone else to this PR?

I am worried any conflict occurs in the case this PR again pends for a long time, since this PR changes a number of files.

**Review by adamleff — APPROVED** [review]

**Comment by jquick** [comment]
Thanks for sticking with us @eramoto !

**Comment by eramoto** [comment]
Thanks for your rapid work @jquick @adamleff !

**Pull request merged**
